[BLOG] PyData Sphinx Theme Accessibility Fixes (#867)
Quansight/Quansight-website: apps/labs/posts/pydata-theme-a11y-fixes.mdx

---
title: 'PyData Theme Accessibility Fixes'
authors: [gabriel-fouasnon]
published: June 26, 2024
description: 'A small showcase of accessibility improvements made to the PyData Sphinx Theme, Fall 2023-Spring 2024'
category: [Access-centered]
featuredImage:
  src: /posts/hello-world-post/featured.png
  alt: 'Excellent alt-text describing the featured image'
hero:
  imageSrc: /posts/hello-world-post/hero.jpeg
  imageAlt: 'Excellent alt-text describing the hero image'
---

# PyData Theme Accessibility Fixes, Before and After

Over the past several months, I have been working with a few coworkers and the
PyData community at large to make some accessibility improvements to the PyData
Sphinx Theme. For this blog post, I chose three such improvements (out of many!)
to examine with a before-and-after lens to give a better idea of how these kinds
of changes can create an improved user experience for disabled people—and often

> for people without disabilities, too.

Reviewer: This is a small nit-pick, so feel free to ignore if you don't feel the same. As a non-native english speaker, I found the sentence _"For this blog post, I chose...without disabilities, too."_ a bit too long and hard to follow. I think breaking it into two sentences could help.

Something like this maybe:

> For this blog post, I chose three such improvements (out of many!) to examine with a before-and-after lens. This will give a better idea of how these kinds of changes can improve user experience for disabled people—and often
> for people without disabilities, too.

For those not familiar with Sphinx, it is a framework for writing documentation.

> It is used, for example, to generate the official Python docs. As with many such

Reviewer 2: docs -> documentation ? in a few places maybe. I thing `docs` is colloquial.

content generation frameworks, Sphinx supports customization via themes. One
such theme, created by and for the PyData community, is the PyData
Theme.

> For those not familiar with accessibility, it is the practice of adapting built

Reviewer 2: 2 paragraph starting with `For those not familiar with`.

> main, Breadcrumb navigation, article, On this page navigation, footer

Still a little bit of cleanup to be done, but much better, right?

![](/posts/pydata-theme-a11y-fixes/landmarks-after.png)

If you’re curious about how we did this, you can check out my [landmarks pull
request](https://github.com/pydata/pydata-sphinx-theme/pull/1454) on GitHub.
Here’s also a short list summarizing what I did:

- Since the theme already had a `<main>` tag, I removed the `role="main"`
  attribute from the `<article>` tag
- Wrapped the header in a `<header>` tag (landmark = "banner")
- Got rid of nested `<nav>`s
- Put the announcement banner in an `<aside>` tag (landmark = "complementary")
- Used `aria-label` and `aria-labelledby` attributes to label some of the
  landmarks (for example, to disambiguate one navigation from another)

Community feedback was key. I did not realize when I removed the `role="main"`
attribute that [Sphinx uses the corresponding
[role="main"]](https://github.com/sphinx-doc/sphinx/blob/4fbd3682d5cdf606159630f7daf6f818d58de05e/sphinx/themes/basic/static/searchtools.js#L185)
CSS attribute selector to highlight search results when a reader searches a

> Sphinx-generated docs site. But one of our user-contributors, Michael Wisely,

Reviewer 3: ```suggestion
Sphinx-generated docs site. But one of our user-contributors, [Michael Wisely][(https://github.com/michael-wisely-gravwell),
```

[reported the missing search
highlight](https://github.com/pydata/pydata-sphinx-theme/issues/1676) on GitHub.

> Then maintainer Daniel McCloy connected the dots. The fix was fortunately

Reviewer 3: ```suggestion
Then maintainer [Daniel McCloy](https://github.com/drammock) connected the dots. The fix was fortunately
```

straightforward. It just required redundantly adding the role attribute to the
main tag, like so: `<main role="main">`. (But don't do that in your own
day-to-day HTML! You can and should just use the `<main>` tag without the role
attribute.)

If you're curious about how I generated the screenshots in the pull request that
I linked, or if you're curious about how your own website looks in terms of
landmarks, you can download a browser extension like
[WAVE](https://wave.webaim.org/extension/) or [Accessibility
Insights](https://accessibilityinsights.io/downloads/). These extensions provide
a way to view all of the landmarks on a web page at once.

## Sidebar Table of Contents, Before and After

> This section assumes some knowledge of HTML and CSS.

Reviewer 3: I liked that you took a moment to provide more, optional information on a screen reader in the last section, so I'm suggest equivalent links in this section. 

```suggestion
This section assumes some knowledge of [HTML](https://en.wikipedia.org/wiki/HTML) and [CSS](https://en.wikipedia.org/wiki/CSS).
```

meaning that a user can both navigate to it using only the keyboard (via the
`Tab` key) **and** a user can "press" the button using only the keyboard (via
`Space`, `Enter` or `Return` keys). But in the case of our foldable/unfoldable
table of contents, it had been implemented without HTML buttons to avoid
JavaScript. It used two tricks in a clever way:

1. In HTML, if you associate a `<label>` with an `<input type="checkbox" />`,
   then clicking the label checks and unchecks the associated checkbox.
2. In CSS, you can select a checked checkbox using the `:checked`
   pseudo-selector.

You can then use CSS to always hide the checkbox and to selectively fold or
unfold the next level in the table of contents based on whether the checkbox is
checked. In code, it looks like the following:

```html
<li>
  <!-- table of contents entry --!>
  <a href="kitchen-sink">Kitchen Sink</a>

  <!-- checkbox hidden via CSS -->
  <input id="toc-checkbox-1" class="toc-checkbox" type="checkbox" />

  <!-- label associated with checkbox -->
  <label for="toc-checkbox-1">🔽</label>

  <!-- next level of TOC -->
  <ul>
    <!-- ... -->
  </ul>
</li>

> ```

Reviewer 2: Not for here, but I'm not sre we should have dark code block for a light theme blog.

Content Accessibility Guidelines success criterion
2.1.1](https://www.w3.org/WAI/WCAG21/Understanding/keyboard.html), this sort of
functionality should be operable by the keyboard in general and not just in the
particular circumstance of having a screen reader.

So what if you turn the label into a tab stop like so: `<label tabindex="0">`,
what do you think happens?

Well, it allows you to reach the element via keyboard, but you still cannot
interact with it by pressing `Enter` or any other key. You cannot check or
uncheck the hidden checkbox this way. So there's no way to make the label
element keyboard-accessible without actually wiring it up with JavaScript to a
keyboard event handler, at which point you might as well just use a button and a
click handler.

This meant that I needed to rethink the way that the table of contents was
implemented.

Some of the ideas that my team came up with:

- Use the [disclosure widget
  pattern](https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/).
- Use a button plus JavaScript
- Use the [tree view
  pattern](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/)

Since I was a little unsure about which direction to take, I decided to ask an

> expert. Here I would like to a give a big shout out to Eric Bailey. Some time

Reviewer 3: And if he is, I'd link…

```suggestion
expert. Here I would like to a give a big shout out to [Eric Bailey](https://ericwbailey.design/). Some time
```

it's best not to use it. For example, the tree view pattern comes with the
expectation that you can expand a node in the tree using the right arrow key and
collapse the same node using the left arrow key, so unless you plan to implement
those keyboard shortcuts with your own JavaScript code, then you shouldn't use
the associated ARIA role.

After the very helpful discussion with Eric Bailey, I decided I would use the
disclosure pattern. This made sense to me for two reasons:

1. The pattern fits the needed UX for a foldable table of contents very well since
   its purpose is to hide and unhide content.
2. It does not require JavaScript and it is available in all modern browsers,
   which increases the chances that it will work well across a broad range of
   devices and assistive technologies.

From there, it was _almost_ a straightforward implementation. 😆

Can you guess what problem I ran into from the following HTML?

```html
<li>
  <details>
    <summary>
      <a href="topics">Topic guides</a>
      <span role="presentation" class="toggle-icon" />
    </summary>
    <ul>
      <!-- next level of toc -->
    </ul>
  </details>
</li>
```

> [TODO: create a Codepen to illustrate this?]

Reviewer 3: I think the one that is there works for my understanding.

The issue is that browsers consider everything in the summary tag to be
clickable content that, when clicked, shows or hides the details. But for our
table of contents, the section headings are also pages that we need to link to.
This makes the click ambiguous. Does the user want to navigate to the opening
page of that section or expand the table of contents at that section? It also
created an ambiguity for keyboard interactivity. With the markup above, when the
user pressed the tab key, the entire section heading plus the expand/collapse
icon is selected (as shown in the following screenshot). The user has to press the
tab key a second time to focus on just the link. So again, it's unclear whether
pressing the `Enter` key should follow the link or expand the section.

![](/posts/pydata-theme-a11y-fixes/ambiguous-summary.png)

> With the help of my designer partner in crime, Smera Goel, we decided that we

Reviewer 3: ```suggestion
With the help of my designer partner in crime, [Smera Goel](https://github.com/smeragoel), we decided that we
```

should present the two elements, the link and the expand/collapse icon, as
completely separate. This led to a new, slightly different markup structure,
which allows the toggle button to be focussed separately from the section link:

```html
<li>
  <a href="topics">Topic guides</a>
  <details>
    <summary>
      <span role="presentation" class="toggle-icon" />
    </summary>
    <ul>
      <!-- next level of toc -->
    </ul>
  </details>
</li>
```

> [TODO: create a Codepen to illustrate this?]

Reviewer 3: Other than the existing one? Or are you asking if the existing one makes sense?

While I can't speak from a regular code-writer's perspective, I think having the information represented as code and rendered supports understanding the how behind the before and afters you are showing. This Codepen is also not very long so I don't think it becomes a big obstacle by accident.

focus indicators. By default, browsers use a focus ring to show focus. This is
what the default focus ring in Chrome looks like:

| Button not focussed (no focus ring)                   | Button focussed (default browser focus ring)                     |
| ----------------------------------------------------- | ---------------------------------------------------------------- |
| ![](/posts/pydata-theme-a11y-fixes/chrome-button.png) | ![](/posts/pydata-theme-a11y-fixes/chrome-button-focus-ring.png) |

Improving and ensuring focus indicators across all interactive elements in the
PyData theme was a major part of the accessibility work we did in the past six
months. There were lots of places where focus rings were missing, incomplete,
low contrast, or just didn't look good. In this section, I've rounded up and
grouped several examples to show-and-tell the before-and-after difference.

### Missing focus rings

The most important kind of before-and-after fix that we made was to add focus
rings to places where they were missing. Now, you might wonder why there were
missing focus rings to begin with, since the browser provides them by default. TODO:

> Ask Chris Holdgraff if he knows about Sphinx-copybutton and Sphinx-togglebutton

Reviewer 2: I looked in git history and found no particular reasons.

Reviewer 2: ping @choldgraf if he remembers why.

the focus ring. It is squished and does not contain the focussed element.

| Gallery tile                                                                                                                             | In-page table of contents                                                                                                  |
| ---------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------- |
| ![Before: broken focus ring appears squished at the bottom of a gallery tile](/posts/pydata-theme-a11y-fixes/focus-ring-tile-before.png) | ![Before: low contrast focus ring in table of contents](/posts/pydata-theme-a11y-fixes/focus-ring-right-toc-before.png)    |
| ![After: fixed focus ring wraps around gallery tile](/posts/pydata-theme-a11y-fixes/focus-ring-tile-after.png)                           | ![After: improved contrast focus ring in table of contents](/posts/pydata-theme-a11y-fixes/focus-ring-right-toc-after.png) |

> ### Frankenstates

Author: Mine too :)

> My colleague Smera Goel, while working through the focus ring designs, coined a

Reviewer: Yes please, go ahead! 😸

There isn't a single GitHub pull request that I can point to in which all of
these focus ring issues were fixed. They were done over many weeks and PRs,
including but not limited to:
[#1415](https://github.com/pydata/pydata-sphinx-theme/pull/1415),
[#1555](https://github.com/pydata/pydata-sphinx-theme/pull/1555),
[#1561](https://github.com/pydata/pydata-sphinx-theme/pull/1561),
[#1564](https://github.com/pydata/pydata-sphinx-theme/pull/1564),
[#1582](https://github.com/pydata/pydata-sphinx-theme/pull/1582),
[#1589](https://github.com/pydata/pydata-sphinx-theme/pull/1589),
[#1778](https://github.com/pydata/pydata-sphinx-theme/pull/1778).

## And more!

This blog post is not meant to capture all of the accessibility work that we did
over the past six months, but rather my hope is that I have given you a taste of
what this work looks like from the developer side of things.

If you would like to take a more exhaustive look, your best option would be to
look at all of the [PST issues and PRs tagged with
accessibility](https://github.com/pydata/pydata-sphinx-theme/issues?q=label%3A%22tag%3A+accessibility%22).
If you find anything you would like to see written about in a blog post like

> this one, please let us know!

Reviewer 3: I would hesitate to simply because I don't know what channels of communication we'll maintain. For example, we don't have an email I'm aware of that we are always watching, and I don't know if authors get notified of blog post comments.

> this one, please let us know!

Reviewer 4: For PST feedback, this should be the PST repo and issues.
We do not have comments in our blog but there is a connect@quansight.com
email that folks generally use to contact us and we get questions redirected (Ralf and I). So we can do something like

"If you find anything you want to see written about in a blog post like
this one, please let us know! You can contact us at connect@quansight.com."

> ## Accessibility in this Blog Post

Reviewer 3: This is a cool section. Even though it felt separate from the rest of the blog post to me, I would keep it. For example, I totally went to check the description for the code and was pleasantly surprised to find Codepens instead of images.

If you write online content, part of your accessibility story will likely be in
what's provided by the authoring tools you choose (like PyData Sphinx Theme),
but some part of it is going to be more directly in your hands, in how you
present and structure your content. There are a number of things that I did in
writing this blog post to make it more accessible, for example:

> - All of the images have alt text or are described by the surrounding text. Even

Reviewer 4: ```suggestion
- All images have alt text or are described by the surrounding text (in which case the alt text is an empty string). Even
```
this is the recommended practice and some folks might not know
